WebDriverAgentLib/Categories/XCUIElement+FBFind.m

```diff
@@ -78,23 +80,62 @@ - (void)descendantsWithProperty:(NSString *)property value:(NSString *)value par
 
 #pragma mark - Search by xpath
 
+- (XCUIElement *)fb_firstDescendantMatchingXPathQuery:(NSString *)xpathQuery
+{
+  // XPath will try to match elements only class name, so requesting elements by XCUIElementTypeAny will not work. We should use '*' instead.
+  xpathQuery = [xpathQuery stringByReplacingOccurrencesOfString:@"XCUIElementTypeAny" withString:@"*"];
+  NSArray *matchingSnapshots = [self.lastSnapshot fb_descendantsMatchingXPathQuery:xpathQuery];
+  if (0 == [matchingSnapshots count]) {
+    return nil;
+  }
+  XCElementSnapshot *firstMatch = [matchingSnapshots firstObject];
+  NSArray *elements = [[self descendantsMatchingType:firstMatch.elementType] allElementsBoundByIndex];
+  NSEnumerator *elementsEnumerator = [elements objectEnumerator];
+  // Speed up elements lookup by starting from the end of the elements array if 'last()' function call is detected in xpath query
+  if ([xpathQuery containsString:@"last()"]) {
+    elementsEnumerator = [elements reverseObjectEnumerator];
+  }
+  for (XCUIElement *element in elementsEnumerator) {
+    if ([element.fb_lastSnapshot _matchesElement:firstMatch]) {
+      return element;
+    }
+  }
+  return nil;
+}
+
 - (NSArray<XCUIElement *> *)fb_descendantsMatchingXPathQuery:(NSString *)xpathQuery
 {
   // XPath will try to match elements only class name, so requesting elements by XCUIElementTypeAny will not work. We should use '*' instead.
   xpathQuery = [xpathQuery stringByReplacingOccurrencesOfString:@"XCUIElementTypeAny" withString:@"*"];
   NSArray *matchingSnapshots = [self.lastSnapshot fb_descendantsMatchingXPathQuery:xpathQuery];
-  NSArray *allElements = [[self descendantsMatchingType:XCUIElementTypeAny] allElementsBoundByIndex];
-  NSArray *matchingElements = [self filterElements:allElements matchingSnapshots:matchingSnapshots];
+  if (0 == [matchingSnapshots count]) {
+    return @[];
+  }
+  // Prefiltering elements speeds up search by XPath a lot, because [element resolve] is the most expensive operation here
+  NSSet *matchedTypes = [XCElementSnapshot fb_getUniqueTypes:matchingSnapshots];
+  NSDictionary *elementsByType = [self splitElementsByType:matchingSnapshots byTypes:matchedTypes];
+  NSArray *matchingElements = [XCUIElement filterElements:elementsByType matchingSnapshots:matchingSnapshots];
   return matchingElements;
 }
 
-- (NSArray<XCUIElement *> *)filterElements:(NSArray<XCUIElement *> *)elements matchingSnapshots:(NSArray<XCElementSnapshot *> *)snapshots
+- (NSDictionary<NSNumber *, NSArray<XCUIElement *> *> *)splitElementsByType:(NSArray<XCElementSnapshot *> *)snapshots byTypes:(NSSet *)types
```

Comment:
I would really consider moving this method to `XCElementSnapshot+FBHelpers.h` so that you can write tests on it.